### PR TITLE
Add instance completion plugin

### PR DESCRIPTION
**Pull request by pschorf**
## Changes proposed in this PR
- Adds a new plugin to be invoked on instance completion 

## Why are we making these changes?
Allows custom integrations to be run on instance completion.

### CODE REVIEW

File: scheduler/src/cook/config.clj [hunk 1]
```diff
@@ -408,17 +408,18 @@
...
+                                  (fn [job-launch-filter]
+                                    (merge
+                                     {:age-out-last-seen-deadline-minutes 10
+                                      :age-out-first-seen-deadline-minutes 600
+                                      :age-out-seen-count 10}
+                                     job-launch-filter)))
+                       (update :job-submission-validator
+                               (fn [job-submission-validator]
+                                 (merge {:batch-timeout-seconds 40}
+                                        job-submission-validator)))))}))
```
Comment (dposada) [inline scheduler/src/cook/config.clj]: Is this change needed?

File: scheduler/src/cook/plugins/completion.clj [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            [cook.plugins.definitions :refer [InstanceCompletionHandler]]
+            [cook.plugins.util]
+            [mount.core :as mount]))
+
+(def no-op
+  (reify InstanceCompletionHandler
+    (on-instance-completion [_ _ _])))
+
+(defn create-default-plugin-object
+  "Returns the configred InstanceCompletionHandler, or a no-op if none is defined."
```
Comment (dposada) [inline scheduler/src/cook/plugins/completion.clj]:
```suggestion
  "Returns the configured InstanceCompletionHandler, or a no-op if none is defined."
```

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 3]
```diff
@@ -895,155 +897,193 @@
         (is (= 1 (count killable)))
         (is (= (-> killable first :db/id) inst-cancelled))))))
 
+(defn make-dummy-status-update
+  [task-id reason state & {:keys [progress] :or {progress nil}}]
+  (let [task {:task-id {:value task-id}
+              :reason reason
+              :state state}]
+    task))
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Do we need the `let`?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 4]
```diff
@@ -895,155 +897,193 @@
...
+                                (on-instance-completion [this job instance]
+                                  (reset! plugin-invocation-atom {:instance instance
+                                                                  :job job})))
+        conn (restore-fresh-database! "datomic:mem://test-instance-completion-plugin")
+        driver (reify msched/SchedulerDriver)
+        driver-atom (atom nil)
+        fenzo (sched/make-fenzo-scheduler driver-atom 1500 nil 0.8)]
+    (with-redefs [completion/plugin plugin-implementation]
+      (testing "Mesos task death"
+        (let [job-id (create-dummy-job conn :user "tsram" :job-state :job.state/running
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/scheduler.clj]: I would favor not using `tsram`

File: scheduler/src/cook/config.clj (hunk 1 above)
Comment (pschorf) [reply scheduler/src/cook/config.clj]: No, I think this was changed while I was working on this PR. Reverted.

File: scheduler/src/cook/plugins/completion.clj (hunk 2 above)
Comment (pschorf) [reply scheduler/src/cook/plugins/completion.clj]: Done

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 3 above)
Comment (pschorf) [reply scheduler/test/cook/test/mesos/scheduler.clj]: No, removed.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 4 above)
Comment (pschorf) [reply scheduler/test/cook/test/mesos/scheduler.clj]: Done
